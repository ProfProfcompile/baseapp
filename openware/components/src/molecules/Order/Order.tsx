import * as React from 'react';
import { OrderForm, TabPanel } from '../';

export type FormType = 'buy' | 'sell';

export type DropdownElem = number | string | React.ReactNode;

export interface OrderProps {
    type: FormType;
    orderType: string | React.ReactNode;
    price: number | string;
    amount: number | string;
}

export type OnSubmitCallback = (order: OrderProps) => void;

export interface OrderComponentProps {
    /**
     * Amount of money in a wallet
     */
    available?: number;
    /**
     * Payment fee that is used in total price calculation
     */
    feeBuy: number;
    /**
     * Payment fee that is used in total price calculation
     */
    feeSell: number;
    /**
     * Callback which is called when a form is submitted
     */
    onSubmit: OnSubmitCallback;
    /**
     * If orderType is 'Limit' this value will be used as price for buy tab
     */
    priceMarketBuy: number;
    /**
     * If orderType is 'Limit' this value will be used as price for sell tab
     */
    priceMarketSell: number;
    /**
     * Name of currency for price field
     */
    from: string;
    /**
     * Name of currency for amount field
     */
    to: string;
    /**
     * Whether order is disabled to execute
     */
    disabled?: boolean;
    handleSendType?: (index: number, label: string) => void;
}

const orderTypes: DropdownElem[] = [
    'Limit',
    'Market',
];

class Order extends React.PureComponent<OrderComponentProps> {
    public render() {
        return (
            <div className="cr-order">
                <TabPanel fixed={true} panels={this.getPanels()} onTabChange={this.handleChangeTab}/>
            </div>
        );
    }

    private getPanels = () => {
        const {
            available,
            disabled,
            feeBuy,
            feeSell,
            priceMarketBuy,
            priceMarketSell,
            from,
            to,
        } = this.props;
        return [
            {
                content: (
                    <OrderForm
                        disabled={disabled}
                        fee={feeBuy}
                        type="buy"
                        from={from}
                        to={to}
                        available={available}
                        priceMarket={priceMarketBuy}
                        onSubmit={this.props.onSubmit}
                        orderTypes={orderTypes}
                    />
                ),
                label: 'Buy',
            },
            {
                content: (
                    <OrderForm
                        fee={feeSell}
                        type="sell"
                        from={from}
                        to={to}
                        available={available}
                        priceMarket={priceMarketSell}
                        onSubmit={this.props.onSubmit}
                        orderTypes={orderTypes}
                    />
                ),
                label: 'Sell',
            },
        ];
    };

    private handleChangeTab = (index: number, label?: string) => {
        if (this.props.handleSendType && label) {
          this.props.handleSendType(index, label);
        }
    }
}

export {
    Order,
    orderTypes,
};
