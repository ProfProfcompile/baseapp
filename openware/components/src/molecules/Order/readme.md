Order example:

```js
const onSubmit = (value) => console.log(value);

<Order
  feeBuy={0.0001}
  feeSell={0.0002}
  onSubmit={onSubmit}
  priceMarketBuy={5}
  priceMarketSell={10}
  from="btc"
  to="eth"
/>
```
