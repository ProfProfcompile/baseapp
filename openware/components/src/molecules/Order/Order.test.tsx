import { shallow, ShallowWrapper } from 'enzyme';
import * as React from 'react';
import { spy } from 'sinon';
import { TabPanel } from '../';
import { Order, OrderProps } from './Order';

const defaultProps = {
    feeBuy: 10,
    feeSell: 10,
    onSubmit: spy(),
    priceMarketBuy: 5,
    priceMarketSell: 10,
    from: 'btc',
    to: 'eth',
};

const setup = (props: Partial<OrderProps> = {}) =>
    shallow(<Order {...{ ...defaultProps, ...props }} />);

describe('Order', () => {
    let wrapper: ShallowWrapper;

    beforeEach(() => {
        wrapper = setup();
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render tabs', () => {
        const tabPanel = wrapper.find(TabPanel);
        expect(tabPanel).toHaveLength(1);
    });
});
