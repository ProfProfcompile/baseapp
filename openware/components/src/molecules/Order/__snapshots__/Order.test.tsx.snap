// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order should match snapshot 1`] = `
<div
  className="cr-order"
>
  <TabPanel
    fixed={true}
    hideMode="hide"
    onTabChange={[Function]}
    panels={
      Array [
        Object {
          "content": <OrderForm
            fee={10}
            from="btc"
            onSubmit={[Function]}
            orderTypes={
              Array [
                "Limit",
                "Market",
              ]
            }
            priceMarket={5}
            to="eth"
            type="buy"
          />,
          "label": "Buy",
        },
        Object {
          "content": <OrderForm
            fee={10}
            from="btc"
            onSubmit={[Function]}
            orderTypes={
              Array [
                "Limit",
                "Market",
              ]
            }
            priceMarket={10}
            to="eth"
            type="sell"
          />,
          "label": "Sell",
        },
      ]
    }
  />
</div>
`;
